Vendoring in libnetwork to 3be488927db8d719568917203deddd630a194564

mavenugo opened the pull request:
This PR brings the vendored `libnetwork` code to 3be488927db8d719568917203deddd630a194564, which pulls in quite a few fixes to support libkv, windows daemon compilation fixes, multi-network support for Bridge driver, etc...

Signed-off-by: Madhu Venugopal madhu@docker.com

icecrime commented:
Wait, this vendors Swarm into Docker, and Swarm already vendors Docker.

![image](https://cloud.githubusercontent.com/assets/1564054/7950661/cc978402-0952-11e5-95af-f8b17c268b70.png)

mavenugo commented:
:) @icecrime `libnetwork` uses swarm packages `discovery` and `store` as libraries and unlike Godeps, I couldn't selectively pick those packages alone in the vendor.sh ...  and hence it shows up like a circular dependancy. Is there a way to just pull in the required packages instead of the entire project ?

icecrime commented:
Intuitively, and without looking into more details, I'm tempted to say that `discovery` and `store` should be pulled out of Swarm as a common dependency to both Swarm and libnetwork, but I'm sure if it was that easy it would already be the case :-/

mavenugo commented:
@icecrime yes, we started to work as an independent project for the store, but decided to pull it into swarm for execution reasons. We can pull it out. am not too sure about `discovery`. @vieux @aluzzardi do you have any comments ?

LK4D4 commented:
It would be cool if we have more nice libraries under docker/ account.

mavenugo commented:
@icecrime thanks to the distribution vendoring process... followed a similar approach of removing unused packages. Now the dependencies are narrowed to just those packages that libnetwork is dependent on (discovery, store, consul/etcd/zookeeper client apis).

tianon commented on hack/vendor.sh:
```diff
@@ -55,9 +55,39 @@ clone hg code.google.com/p/go.net 84a4013f96e0
+mkdir -p src/github.com/hashicorp/consul
+mv tmp-api src/github.com/hashicorp/consul/api
+
+clone git github.com/coreos/go-etcd 73a8ef737e8ea002281a28b4cb92a1de121ad4c6
```
Can we use a tagged release of `go-etcd` instead? (https://github.com/coreos/go-etcd/releases)

Same with consul (https://github.com/hashicorp/consul/releases).

tianon replied:
(and if they don't have a recent enough release, we've found that most upstreams are pretty reasonable about tagging a new one)

LK4D4 replied:
looks like we need func for such vendoring :)

mavenugo replied:
@tianon good point. I will cross-check with swarm discovery/store to make sure the selected tag satisfies the dependency needs.

LK4D4 commented on hack/vendor.sh:
```diff
@@ -55,9 +55,39 @@ clone hg code.google.com/p/go.net 84a4013f96e0
+clone git github.com/deckarep/golang-set ef32fa3046d9f249d399f98ebaf9be944430fd1d
+
+#libnetwork uses swarm discovery and store libraries
+clone git github.com/docker/swarm 54dfabd2521314de1c5b036f6c609efbe09df4ea
```
looks like we need func for such stuff :)

mavenugo replied:
@LK4D4 @tianon if we are okay with this approach, then yes, having a generic function to filter out unused packages & pulling in only the necessary packages will be helpful.

abronan commented:
@mavenugo @icecrime We will pull `pkg/store` out in its own repository but not sure this is going to happen before 1.7.0. If you found a way to vendor only the relevant parts then it's all good.

We can try to pull out `discovery` the same way if this is more convenient for `libnetwork` (we'd like to do this in Swarm as well I guess but this is a bit more difficult as it is more tied with Swarm).

icecrime commented:
* I don't think we have any dependency on 1.7.0 there
* I'm +1 for pulling out rather than import from swarm, but that's just my 2cts

calavera commented:
Do we have any resolution to move this forward?

@icecrime the vendor script actually removes all the pieces that we don't need from swarm. I agree that's not ideal, but it might not be too bad in the short term.

@abronan is there anything I can help to extract that discovery package? Is it just a matter of extracting that directory to a new repo and updating dependencies, or we should take something else into account?

icecrime commented:
@calavera I think it has been taken care of (circular dependency was broken), but that the  PR was not yet updated.

mavenugo commented:
Yes. I will push an update today.

mavenugo committed:
Vendoring in libnetwork to 3be488927db8d719568917203deddd630a194564

This PR brings the vendored libnetwork code to
3be488927db8d719568917203deddd630a194564, which pulls in quite a few
fixes to support kvstore, windows daemon compilation fixes,
multi-network support for Bridge driver, etc...

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo commented:
@icecrime @calavera @abronan migrated to libkv and addressed other comments. PTAL.

mavenugo commented:
@jhowardmsft FYI. pls check & confirm if the windows compilation issues related to libnetwork are addressed.

aluzzardi commented:
This doesn't load on my browser :) Just to check, are you now referencing to `docker/libkv` instead of `docker/swarm/pkg/store`? We put it in a separate repo a few days ago

mavenugo commented:
@aluzzardi yes. thanks for making that happen. appreciate it.

calavera commented:
It looks like we're vendoring a testing library that we don't need. I think https://github.com/docker/libkv/pull/6 should fix this problem.

mavenugo commented:
@calavera yes.

jhowardmsft commented:
@mavenugo  ![lgtm](https://cloud.githubusercontent.com/assets/10522484/8140203/4d52445c-110e-11e5-8f99-4d9be9f1c1c2.JPG)
LGTM (from a Windows compile point of view) :+1:

mavenugo commented:
Thanks @jhowardmsft

tianon commented on hack/vendor.sh:
```diff
@@ -17,9 +17,16 @@ clone hg code.google.com/p/go.net 84a4013f96e0
-clone git github.com/docker/libnetwork e578e95aa101441481411ff1d620f343895f24fe
+clone git github.com/docker/libnetwork 3be488927db8d719568917203deddd630a194564
+clone git github.com/docker/libkv ab16c3d4a8785a9877c62d0b11ea4441cf09120c
+clone git github.com/stretchr/testify dab07ac62d4905d3e48d17dc549c684ac3b7c15a
```
Can we use `v1.0` here instead? (https://github.com/stretchr/testify/releases)

mavenugo commented:
Thanks @calavera I will pick up a8d22fa21a3b4626aa33364ca2603c84f2622c32 from libkv.

calavera commented:
@mavenugo sorry, you need docker/libkv#7. There was some confusion about the right fix.

mavenugo commented:
@calavera @abronan its a good fix. but unfortunately libkv uses the mock package from testify that pulls in the deps back in. seems like mock is a service provided by libkv. :(

mavenugo commented:
Thanks @abronan @aluzzardi for offering to remove the mock from libkv top level provider. That will surely get rid of testify.

mavenugo commented:
@aluzzardi @abronan Thanks for the quick work. testify is now removed.
@calavera @icecrime PTAL

icecrime commented:
Ping @tianon.

tianon commented:
LGTM :+1:

tianon commented:
I'm very curious what libnetwork uses TOML internally for though. :smile:

mavenugo commented:
@tianon Thanks for the review. libnetwork provides a tool named dnet, which uses TOML for configuration.

icecrime commented:
LGTM! Thanks for extracting the libs <3
